x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/package-info.java

```diff
@@ -99,31 +99,81 @@
+ *      the rest of the planner steps will act on.  Generally speaking, things which change the grammar will need to make changes here
+ *      as well.</li>
+ *      <li>{@link org.elasticsearch.xpack.esql.analysis.PreAnalyzer} - Acts somewhat like a linker, to find the indices and enrich
+ *      policies referenced by the query.</li>
+ *     <li>{@link org.elasticsearch.xpack.esql.analysis.Analyzer} - Resolves references (such as variable names) and performs implicit
+ *     casting.</li>
+ *     <li>{@link org.elasticsearch.xpack.esql.analysis.Verifier} - Responsible for failing impossible queries with a human readable error
+ *     message.</li>
+ *     <li>{@link org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizer} - Global optimizations based strictly on the structure of
+ *     the query (i.e. not factoring in information about the backing indices).  The bulk of query transformations happen in this step.
+ *     This has two important sub-phases:
+ *     <ul>
+ *         <li>The Substitution phase rewrites things to expand out shorthand in the syntax.  For example, an eval embedded in a
+ *         stats gets replaced with an eval followed by a stats.  This phase also applies aggregation surrogates, such as replacing
+ *         an average with a sum and a count.</li>
+ *         <li>"Operator Optimization" (NB: The word "operator" is extremely overloaded and referrers to many different things.) transform
+ *         the tree in various different ways.  This includes folding (i.e. computing constant expressions at parse time), combining
+ *         expressions, dropping redundant clauses, and some normalization such as putting literals on the right whenever possible.  These
+ *         rules are run in a loop until none of the rules make any changes to the plan (there is also a safety shut off at 100
+ *         iterations, although we expect to never hit that)</li>
+ *     </ul>
+ *     </li>
+ *     <li>{@link org.elasticsearch.xpack.esql.planner.Mapper} - Translates the logical plan into a physical plan.  This is where we start
+ *     to decide what will be executed on the data nodes and what will be executed on the coordinator nodes.  This step creates
+ *     {@link org.elasticsearch.xpack.esql.plan.physical.FragmentExec} instances, which represent logical plan fragments to be sent
+ *     to the data nodes and {@link org.elasticsearch.xpack.esql.plan.physical.ExchangeExec} instances, which represent data being
+ *     sent back from the data nodes to the coordinating node.</li>
+ *     <li>{@link org.elasticsearch.xpack.esql.optimizer.PhysicalPlanOptimizer} - Apply any global physical optimizations, such as
+ *     projecting away columns.</li>
+ *     <li>{@link org.elasticsearch.xpack.esql.optimizer.LocalLogicalPlanOptimizer} - Data node level logical optimizations.  At this point
+ *     we have access to {@link org.elasticsearch.xpack.esql.stats.SearchStats} which provides access to metadata about the index</li>
```

Comment:
I believe this one includes all of the steps that are part of the `LogicalPlanOptimizer` too.

Reply:
It does? Where does that happen?

Reply:
https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java#L81